sklearn/utils/estimator_checks.py

```diff
@@ -1443,41 +1443,90 @@ def check_supervised_y_2d(name, estimator_orig):
     assert_allclose(y_pred.ravel(), y_pred_2d.ravel())
 
 
-@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+@ignore_warnings
+def check_classifiers_predictions(X, y, name, classifier_orig):
+    classes = np.unique(y)
+    classifier = clone(classifier_orig)
+    if name == 'BernoulliNB':
+        X = X > X.mean()
+    set_random_state(classifier)
+
+    classifier.fit(X, y)
+    y_pred = classifier.predict(X)
+
+    if hasattr(classifier, "decision_function"):
+        decision = classifier.decision_function(X)
+        n_samples, n_features = X.shape
+        assert(isinstance(decision, np.ndarray))
+        if len(classes) == 2:
+            dec_pred = (decision.ravel() > 0).astype(np.int)
+            dec_exp = classifier.classes_[dec_pred]
+            assert_array_equal(dec_exp, y_pred,
+                               err_msg="decision_function does not match "
+                               "classifier for %r: expected '%s', got '%s'" %
+                               (classifier, ", ".join(map(str, dec_exp)),
+                                ", ".join(map(str, y_pred))))
+        elif getattr(classifier, 'decision_function_shape', 'ovr') == 'ovr':
+            decision_y = np.argmax(decision, axis=1).astype(int)
+            y_exp = classifier.classes_[decision_y]
+            assert_array_equal(y_exp, y_pred,
+                               err_msg="decision_function does not match "
+                               "classifier for %r: expected '%s', got '%s'" %
+                               (classifier, ", ".join(map(str, y_exp)),
+                                ", ".join(map(str, y_pred))))
+
+    # training set performance
+    if name != "ComplementNB":
+        # This is a pathological data set for ComplementNB.
```

Comment:
Could you explain what is happening there?

Reply:
@glemaitre , this code already existed from the multiclass case. I moved it to a separate method in order to be able to use it both for multiclass and binary case. I can try to investigate .

Reply:
It seems that the `ComplementNB` has been added recently. During prediction it doesn't always predict all possible classes. For example: the input contains 3 unique classes but the prediction has only 2 unique classes. I have to read to implementation in order to understand why this happens.

Reply:
@jnothman , Do you know why this happens ? Is this a feature of ComplementNB ?

Reply:
@airalcorn2 might have an idea...

Reply:
It's really just how the math shakes out for this particular data set. The complement feature counts for the three classes are:

```python
[[ 28.19448326  20.20988056]
 [ 21.56648936  44.53201548]
 [ 45.26121905  30.28029562]]
```

so the complement count totals for each class are:

```python
[[ 48.40436382]
 [ 66.09850484]
 [ 75.54151467]]
```

As a result, the normalized complement features are:

```python
[[ 0.58247813  0.41752187]
 [ 0.32627802  0.67372198]
 [ 0.59915689  0.40084311]]
```

As you can see, the normalized complement feature vectors are very similar for the 0 and 2 classes. The final `feature_log_prob_` matrix (so, after taking the negative log of the above) looks (when transposed) like:

```python
[[ 0.54046364  1.12000544  0.51223179]
 [ 0.87341835  0.39493775  0.91418518]]
```

The features for the 0 class in the data matrix are the following:

```python
[[ 1.86111069  2.78745871]
 [ 1.94936262  2.66852288]
 [ 2.0000686   2.57898891]
 [ 1.90144365  2.69194788]
 [ 1.68480541  2.73007368]
 [ 1.87215982  2.62201183]
 [ 1.84494713  2.71265579]
 [ 1.90820282  2.6300342 ]
 [ 1.88070993  2.70714243]
 [ 1.91380191  2.67237895]]
```

So the second feature is consistently ~1.4 times larger than the first feature for this particular class. Looking back at `feature_log_prob_`, we can see that class 2 has weights of `[0.512, 0.914]` while class 1 has weights of `[0.540, 0.873]`, i.e., class 2 has a larger weight on the second feature than class 1, but only a slightly smaller weight on the first feature. As a result, `ComplementNB` classifies all of the 0 samples as 2.

I think it's just a case of the data not resembling text data, so the assumptions that are baked into `ComplementNB` end up being detrimental.

Full code to explore is below:

```python
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

from sklearn.datasets import make_blobs
from sklearn.naive_bayes import ComplementNB
from sklearn.preprocessing import StandardScaler
from sklearn.utils import shuffle
from sklearn.utils.testing import assert_array_equal

X_multiclass, y_multiclass = make_blobs(n_samples = 30, random_state = 0,
                                        cluster_std = 0.1)
X_multiclass, y_multiclass = shuffle(X_multiclass, y_multiclass,
                                     random_state = 7)
X_multiclass = StandardScaler().fit_transform(X_multiclass)
X_multiclass -= X_multiclass.min() - .1
labels_multiclass = ["one", "two", "three"]
y_names_multiclass = np.take(labels_multiclass, y_multiclass)

(X, y) = (X_multiclass, y_multiclass)

df = pd.DataFrame(dict(x = X[:, 0], y = X[:, 1], color = y))
sns.lmplot("x", "y", data = df, hue = "color", fit_reg = False)
plt.show()

classifier = ComplementNB()
classifier.fit(X, y)
y_pred = classifier.predict(X)
assert_array_equal(np.unique(y), np.unique(y_pred))

print(classifier.feature_count_)
print(classifier.class_count_)
print(classifier.feature_all_)

comp_count = classifier.feature_all_ + classifier.alpha - classifier.feature_count_
print(comp_count)
sums = comp_count.sum(axis = 1, keepdims = True)
print(sums)
normed = comp_count / sums
print(normed)
logged = -np.log(normed)
print(logged.T)

jll = np.dot(X, logged.T)
print(jll)
zero_class = X[np.where(y == 0)]
print(zero_class)
print(zero_class[:, 1] / zero_class[:, 0])
print(jll[np.where(y == 0)])
```

Reply:
Thank you for the detailed explanation @airalcorn2 :)